Support making standalone container and node admin RPMs

hakonhall opened the pull request.

hakonhall committed:
Support making standalone container and node admin RPMs

hakonhall committed:
Move services.xml to application directory and remove stray dependency

hakonhall committed:
Highlight commonality among dependency jar files

hakonhall commented:
Support for:
 - Making standalone JDisc container RPM, including script to start (and eventually stop) based on start-configserver.
 - Making standalone Node Admin RPM, including script to start (and eventually stop)

@aressem for RPM-related changes, e.g. *.spec files.
@arnej27959 for jdisc-container and node-admin scripts.
@gjoranv for the list of dependencies in vespa-jdisc-container.spec. It would be nice to somehow get this list of dependencies from an authoritative source, as discussed. Alternatively, all target/*-jar-with-dependencies.jar and target/dependency/*.jar files could be included, but it would raise the size of the RPM from about 75M to 325M, which I would like to avoid for now. I'm thinking a simplification can be done later if it actually is needed.

FYI @smorgrav

aressem reviewed (commented):
Looks good. The build-rpm.sh script should be placed somewhere else than the root. Probably dist/ would be the best option.

hakonhall committed:
Move build-rpm to dist

gjoranv commented on standalone-container/vespa-jdisc-container.spec:
```diff
@@ -0,0 +1,93 @@
+  cp "$dir"/* "$jars_dir"
+done
+
+declare -a modules=(
```
It seems that you have missed all the jersey2 bundles listed in container-disc/pom.xml.

hakonhall replied:
That's actually taken care of 6 lines above: I'm taking all vespa_jersey2 dependencies (which gets me a few more than I actually need).

gjoranv replied:
ok, good!

hakonhall commented:
ping @arnej27959

arnej27959 commented on node-admin/src/main/sh/node-admin:
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
```
shell scripts should be named '.sh' in the repository (so node-admin.sh)

arnej27959 commented on standalone-container/src/main/sh/jdisc-container:
```diff
@@ -0,0 +1,218 @@
+
+    local vespa_log="$VESPA_HOME/logs/vespa/vespa.log"
+    export VESPA_LOG_TARGET="file:$vespa_log"
+    mkdir -p "$(dirname "$vespa_log")"
```
should use FixDataDirectory here

arnej27959 commented on standalone-container/src/main/sh/jdisc-container:
```diff
@@ -0,0 +1,218 @@
+
+    export VESPA_LOG_CONTROL_FILE="$VESPA_HOME/var/db/vespa/logcontrol/$service.logcontrol"
+    export VESPA_LOG_CONTROL_DIR="$(dirname "$VESPA_LOG_CONTROL_FILE")"
+    mkdir -p "$VESPA_LOG_CONTROL_DIR"
```
should use FixDataDirectory here

arnej27959 commented on standalone-container/src/main/sh/jdisc-container:
```diff
@@ -0,0 +1,218 @@
+#!/bin/bash
```
this script should also be named jdisc-container.sh in the repository

arnej27959 commented on standalone-container/src/main/sh/jdisc-container:
```diff
@@ -0,0 +1,218 @@
+    printenv > "$cfpfile"
+    FixDataDirectory "$bundlecachedir"
+
+    java \
```
should do "exec java" here

arnej27959 commented on standalone-container/vespa-jdisc-container.spec:
```diff
@@ -0,0 +1,93 @@
+chmod +x /etc/profile.d/vespa.sh
+
+%postun
+rm -f /etc/profile.d/vespa.sh
```
only do on uninstall (see https://github.com/vespa-engine/vespa/pull/4347)

aressem replied:
This is also incompatible with the regular vespa rpm because it will have conflicting files and remove the vespa user on upgrade / removal.

arnej27959 commented on standalone-container/src/main/sh/jdisc-container:
```diff
@@ -0,0 +1,218 @@
+	Fail "Too many arguments"
+    fi
+
+    Fail "Not implemented"
```
you can save the pid in a pidfile on start and kill it here in stop

arnej27959 reviewed (commented):
some comments from me on the scripts.

arnej27959 commented:
i'm also worried about you setting VESPA_USER.  is node-admin going to control docker containers including starting stuff inside them? docker-api has the "executeInContainer" method which uses getDefaults().vespaUser() which again uses VESPA_USER.

hakonhall committed:
Fix review comments
 - Rename *jdisc-container* to *standalone-container* (for files in this PR)
 - Implement stop standalone-container command
 - Avoid e.g. vespa user deletion on upgrade

hakonhall reviewed (commented):
Good point on VESPA_USER - I have removed the override. Node admin doesn't currently use VESPA_USER, but as you point out it should and I'll fix that later.

I have also renamed 'jdisc-container' to 'standalone-container' since that seems a bit less ambiguous: sometimes jdisc non-standalone-container is described as 'jdisc container'.

arnej27959 commented:
looks good to me now

hakonhall commented:
@aressem I'll assume you're fine with my response to your comment

aressem commented:
👍
